### ISSUE TITLE
Cite and discuss LeAR ( https://aclanthology.org/2021.findings-acl.97.pdf ) as an example of a tree-structured model that produces a tree-structured syntax representation which does very well on COGS (COGS authors argued such an approach is likely required for structural generalization success)

### ISSUE THREAD

**Issue by willy-b** [willy-b/RASP-for-ReCOGS#86]
As an example of both a tree-structured model and a tree-structured representation (the latter does not necessarily require the former), LeAR  ( https://aclanthology.org/2021.findings-acl.97.pdf ) uses two modules, one that "learns to model the latent syntactic algebra [using a Tree-LSTM to] produce the latent syntactic structure of each expression" and a second that "learns to assign semantic operations to syntactic operations" so that they can "transform a syntactic tree to the final composed semantic meaning", and in this way achieves 97.7% on COGS. However this is neither a Transformer architecture nor a vanilla unstructured neural network, it uses an LSTM specifically architected based on the assumption of a recursive, tree-structured context-free grammar. Note LeAR also uses what appear to be heavily crafted semantic primitives in their Interpreter module, see Table 8 for COGS, requiring manual task specific design.

We show that by using the power of the Decoder loop to unroll recursive aspects of the grammar, a vanilla Transformer with NEITHER a tree-structuring to the model (not using the Transformer to recursively combine pairs of tokens) NOR a tree-structured representation (fixed layer depth for all recursion depths), we can solve ReCOGS (and COGS, see https://github.com/willy-b/RASP-for-COGS ) and we show that Wu et al 2023's trained from scratch Transformer is learning such a non-hierarchical/non-tree structured representation (as opposed to learning a tree-structured representation without being a tree-structured model to bias this, which it could have done, or at least a deeper one could have done) but not succeeding at least in part due not to learning the pp masking rule that we argue is potentially learnable from the center-embedded pp examples in the training set.
